File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
import java.util.*;
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max=0,curr=0;
        HashMap<Character, Integer> map = new HashMap<>();
      for(int i=0;i<s.length();i++){
        char ch = s.charAt(i);
        if(!map.containsKey(ch)){
          map.put(ch, i);
          curr++;
        }else{
          int num = i - map.get(ch);
          if(num <= curr + 1) curr = num;
          else curr++;
          map.put(ch, i);
        }
        max = Math.max(curr, max);
      }
      return max;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
      if(s.length() == 0)
        return 0;
      map<char, int>idx;
      int l = 0;
      idx[s[0]] = 1;
      int ans = 1;
      for(int r = 1; r<s.length(); r++){
        if(idx[s[r]]-1 >= l)
          l = idx[s[r]];
        ans = max(ans, r-l+1);
        idx[s[r]] = r+1;
      } 
      return ans;
    }
};